File: Hyunsookim92/42.TrappingRainWater.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        cnt = 0     # 가둘 수 있는 물의 양
        for i in range(1,max(height)+1):
            tmp = []
            for j in range(len(height)):
                if height[j] >= i:
                    tmp.append(j)

            if min(tmp) != max(tmp):
                for k in range(min(tmp), max(tmp)):
                    if height[k] < i:
                        cnt += 1
        return cnt

height1 = [0,1,0,2,1,0,1,3,2,1,2,1]
height2 = [2,1,0,2,1,0,1,3,2,1,2,1]
height3 = [4,2,0,3,2,5]
height4 = [5,2,0,3,2,5]

print(Solution().trap(height1))

File: t5271688/42.Trapping Rain Water.cpp
class Solution
{
public:
    int trap(vector<int> &height)
    {
        int result = 0;
        int length = height.size();
        int leftMax[100010] = {0,};
        int rightMax[100010] = {0,};
        leftMax[0] = height[0];
        rightMax[length - 1] = height[length - 1];
        for (int i = 1; i < length; ++i)
        {
            leftMax[i] = max(leftMax[i - 1], height[i]);
            rightMax[length - i - 1] = max(rightMax[length - i], height[length - i - 1]);
        }
        for (int i = 0; i < length; i++)
        {
            int val = min(leftMax[i], rightMax[i]);
            result += val > height[i] ? (val - height[i]) : 0;
        }
        return result;
    }
};